components/external_storage/export/src/dylib.rs

```diff
@@ -0,0 +1,308 @@
+                    Err(extern_to_io_err(e))
+                } else {
+                    Ok(())
+                }
+            }
+        }
+    }
+}
+
+impl ExternalStorage for ExternalStorageClient {
+    fn name(&self) -> &'static str {
+        self.blob_storage.name()
+    }
+
+    fn url(&self) -> io::Result<url::Url> {
+        self.blob_storage.url()
+    }
+
+    fn write(
+        &self,
+        name: &str,
+        reader: Box<dyn AsyncRead + Send + Unpin>,
+        content_length: u64,
+    ) -> io::Result<()> {
+        info!("external storage writing");
+        (|| -> anyhow::Result<()> {
+            let file_path = file_name_for_write(&*self.blob_storage, &name);
+            let req = write_sender(
+                &self.runtime,
+                self.backend.clone(),
+                file_path.clone(),
+                name,
+                reader,
+                content_length,
+            )?;
+            let bytes = req.write_to_bytes()?;
+            info!("write request");
+            self.external_storage_write(bytes)?;
+            DropPath(file_path);
+            Ok(())
+        })()
+        .context("external storage write")
+        .map_err(anyhow_to_io_log_error)
+    }
+
+    fn read(&self, _name: &str) -> Box<dyn AsyncRead + Unpin> {
+        unimplemented!("use restore instead of read")
+    }
+
+    fn restore(
+        &self,
+        storage_name: &str,
+        restore_name: std::path::PathBuf,
+        expected_length: u64,
+        speed_limiter: &Limiter,
+    ) -> io::Result<()> {
+        info!("external storage restore");
+        let req = restore_sender(
+            self.backend.clone(),
+            storage_name,
+            restore_name,
+            expected_length,
+            speed_limiter,
+        )?;
+        let bytes = req.write_to_bytes()?;
+        self.external_storage_restore(bytes)
+    }
+}
+
+fn extern_to_io_err(e: ffi_support::ExternError) -> io::Error {
+    io::Error::new(io::ErrorKind::Other, format!("{:?}", e))
+}
+
+type FfiInitFn<'a> =
+    libloading::Symbol<'a, unsafe extern "C" fn(error: &mut ffi_support::ExternError) -> ()>;
+type FfiFn<'a> = libloading::Symbol<
+    'a,
+    unsafe extern "C" fn(error: &mut ffi_support::ExternError, bytes: Vec<u8>) -> (),
+>;
+
+fn external_storage_init_ffi_dynamic(library: &libloading::Library) -> io::Result<()> {
+    let mut e = ffi_support::ExternError::default();
+    unsafe {
+        let func: FfiInitFn = library
+            .get(b"external_storage_init")
+            .map_err(libloading_err_to_io)?;
+        func(&mut e);
+    }
+    if e.get_code() != ffi_support::ErrorCode::SUCCESS {
+        Err(extern_to_io_err(e))?;
+    }
+    Ok(())
+}
+
+fn call_ffi_dynamic(
+    library: &libloading::Library,
+    fn_name: &[u8],
+    bytes: Vec<u8>,
+) -> io::Result<()> {
+    let mut e = ffi_support::ExternError::default();
+    unsafe {
+        let func: FfiFn = library.get(fn_name).map_err(libloading_err_to_io)?;
```

Comment:
I suspect the performance is not noticeable. Currently caching wouldn't do anything because a new ExternalStorageClient is created for each SST file so the functions are only called once.